### DISCUSSION TITLE
Job always a new instance (repro included)

### DISCUSSION THREAD

**Discussion by munkii** [shinyorg/shiny#1509] [Q&A]
We have some mostly static data that I want to download from our server and store in our app local db. A Shiny Job sounds like the right tool.

We are currently adding this to our Xamarin Forms app using Shiny 3.2.4 and XF 5.0.0.2662.

With RunOnForeground defaulting to false I went with that:

- On Android it runs once at startup then never again or at least not that I can see
- On iOS it does not run on startup but does run again but it's not clear when it will run (which would be fine)

I know the Job is being registered because I can do the following

```
var jobManager = Shiny.Hosting.Host.Current.Services.GetRequiredService<IJobManager>();

jobManager.RequestAccess();

var job = jobManager.GetJob(nameof(DownloadClinicsJob));

jobManager.Run(nameof(DownloadClinicsJob));
```

If is set RunOnForeground to true:

- On Android it runs every 30 seconds whether the app is in the Foreground or not
- On iOS it runs every 30 seconds it will run when the app is in the foreground but does not see to run when the app is backgrounded

The background job in the repro project just sends a local notifcation. The ultimate goal is to make an anonymous web requests and write  to a local db.

As per the Shiny docs (for MAUI) and the Shiny source I wanted to stash the run time of the Job and then refer to it at its next run time. For now I wanted to display that run time in the notification. Later I would use it to just ignore runs that come too soon after last. However the last run time is always Null and I'm assuming that is because each time the job runs it is a new instance rather than a singleton.

Have I registered it incorrectly?

I have copied some of the setup from the AppBuilder which might hjave been a mistake because that is understandibly targetting the Shiny MAUI libraries

Repro
https://github.com/munkii/ShinyJobsIssue

**Comment by aritchie** [comment]
There is weird submodule in your repo so I can see the code.  Jobs are transient, however, they work with a concept in Shiny called stateful services.  If you make you job an INPC like a viewmodel, those values will automatically be restored each time the job is run.  

You can take a look here: https://github.com/shinyorg/shiny/blob/dev/src/Shiny.Jobs/Job.cs  <= you can also just inherit this.  It has a LastRunTime property.  Is this what you're using?

**Reply by munkii** [reply]
I am using Shiny.Jobs.Job although I had my own lastRunTime and executionCount fields (not properties).

> There is weird submodule in your repo so I can see the code. 

I made a mess of creating that repo. I will delete it and start from scratch

**Reply by munkii** [reply]
I've updated the repro link. I think it was a lot to do with me having my own lastRunTime. Seems to be a lot better now.

1. What is RunOnForeground for?
2. Should I expect jobs to always run once on startup? (That's what I'd prefer FWIW)

**Reply by aritchie** [reply]
1. RunOnForeground runs the job when your app is in the foreground.  It runs it through a timer which is defaulted to 30 seconds
2. Yes if RunOnForeground is set to true
